File: BOJ/1052.cpp
#include <iostream>

using namespace std;

int main(void) {
	int
		n_bought_bottles = 0,
		N, K;

	// set N and K
	cin >> N >> K;

	while (true) {
		int
			n_bottles = 0,
			sizeof_smallest_bottle = -1;

		for (int i = 0; i < 32; i++) {
			if ((N >> i) & 1) {
				// count the number of bottles
				n_bottles++;

				// get the size of the smallest bottle
				if (sizeof_smallest_bottle == -1) {
					sizeof_smallest_bottle = 1 << i;
				}
			}
		}

		// break the loop when the number of bottles is equals to or lower than {K}
		if (n_bottles <= K)
			break;

		// buy bottles
		n_bought_bottles += sizeof_smallest_bottle;
		N += sizeof_smallest_bottle;
	}

	// print the answer
	cout << n_bought_bottles;

	return 0;
}

File: BOJ/1052.ts
import * as fs from 'fs';

const inputManager: Generator<string, any, unknown> = (function *() {
  const stdin = `${__dirname}/input.txt`;
  // const stdin = '/dev/stdin';
  const lines = fs.readFileSync(stdin).toString().split('\n');
  for (const line of lines) yield line;
  return null;
})();
const nextLine = (): string => inputManager.next().value;

// set N and K
let [N, K] = nextLine().split(' ').map(n => Number.parseInt(n));
let nBoughtBottles = 0;

while (true) {
  // convert N to binary string
  const binaryString = N.toString(2);

  // count the number of bottles
  const nBottles = Array.from(binaryString).filter(bit => bit === '1').length;

  // break the loop when the number of bottles is equals to or lower than {K}
  if (nBottles <= K) break;

  // get the size of the smallest bottle
  const sizeOfSmallestBottle = Number.parseInt(binaryString.substring(binaryString.lastIndexOf('1')), 2);

  // buy bottles
  nBoughtBottles += sizeOfSmallestBottle;
  N += sizeOfSmallestBottle;
}

// print the answer
console.log(nBoughtBottles);
